### ISSUE TITLE
Allow remote FHIR search through FHIR engine

### ISSUE THREAD

**Issue by joiskash** [google/android-fhir#1912]
**Is your feature request related to a problem? Please describe.**
Currently FHIR engine provides sync APIs to periodically fetch data. This is useful for offline systems and data that maybe required after a long period of time (eg: 24 hours). The FhirSyncWorker also has a one time work request to fetch data just in time. The developer has to pass the search URL to the work manager which will run a job to fetch data based on the search URL and save it in the DB.
For an application to fetch a Patient resource that is not synced yet we need to do the following:

1. Search in local DB using fhirEngine
2. If not found, create a one time work request
3. After the completion of the task, wait for indexing to complete
4. Search again in the local DB using fhirEngine.

**Describe the solution you'd like**
There be value in automatically searching for the resource in the Remote server if not found and internet is available. That way we avoid the multiple step process mentioned above.

**Describe alternatives you've considered**

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

cc: @f-odhiambo @omarismail94

**Comment by jingtang10** [comment]
thanks for raising this @joiskash. qq: would it help to increase the periodic sync frequency?

the idea of the offline sync api + search api is really that the app shouldn't be trying to handle online/offline all the time. the app should treat on-device data as if it's online all the time.

**Comment by joiskash** [comment]
The expectation of users is that if the device is online, then it should fetch the latest data. We use online server queries to fetch list of patients that have visited the clinic on a given day.
some eg:
```
 Encounter?service-provider=faciliityid&
   class=IMP&
   date=2023-05-12&
  _sort:desc=_lastUpdated&
  _include=Encounter:patient
```

On clicking the Patient, we use a normal sync job. The Online FHIR search is mainly to fetch some resources for display. The actual downloading and inserting in DB is all done by the SDK.

**Comment by fredhersch** [comment]
@joiskash Is this still an open issue or have you implemented this into your client application? Can you provide any updates. Thanks

**Comment by jingtang10** [comment]
I will close this but for the record I want to add some notes:

[HAPI's android client](https://hapifhir.io/hapi-fhir/docs/android/client.html) actually supports online search I believe. It just downloads the fhir resources but doesn't persist them. If the app wants just to do that it could either use the HAPI client or do something very similar.

The value of the FHIR Engine library is that it keeps copies of FHIR resources and handles syncing with the server. The set of resources to be downloaded to FHIR Engine from the backend is effectively defined by the implementation of [DownloadWorkManager](https://github.com/google/android-fhir/blob/master/engine/src/main/java/com/google/android/fhir/sync/DownloadWorkManager.kt). And then, any resources created locally is added to the set and uploade to the backend.

So if we create a new online search API, we'd need to clarify what it means to the download work manager... does it mean that the download work manager now need to keep the newly downloaded resources from the online search API in sync with the server in the future? If that's the case, then it feels like we should just use a one time sync. If not, then I don't know why we need to keep these resources in the db.

To summarise, it's a different approaching of handling resources altogether... Without better understanding of the user cases I hesitate to just create a new API like that. I think we should keep our eyes open for similar feature requests and reopen this if needed.
